File: sdk-go/compose.go
package composr

import (
	"encoding/json"
	"fmt"
	"math"
	"math/rand"
	"strings"
	"time"
)

// sdkVersion is the version of this Go SDK.
const sdkVersion = "0.1.0"

// compose assembles a prompt locally from cached config.
func compose(config *SDKConfig, compositionName string, ctx ComposeContext, opts *composeOptions) (*ComposeResult, error) {
	// Find composition by name
	var comp *CompositionConfig
	for i := range config.Compositions {
		if config.Compositions[i].Name == compositionName {
			comp = &config.Compositions[i]
			break
		}
	}
	if comp == nil {
		return nil, fmt.Errorf("composition %q not found", compositionName)
	}

	nodes := comp.Graph.Nodes
	edges := comp.Graph.Edges

	// Build lookup maps
	nodeMap := make(map[string]*GraphNode, len(nodes))
	for i := range nodes {
		nodeMap[nodes[i].ID] = &nodes[i]
	}

	edgesBySource := make(map[string][]GraphEdge, len(edges))
	for _, edge := range edges {
		edgesBySource[edge.Source] = append(edgesBySource[edge.Source], edge)
	}

	// Enrich context with auto-captured metadata
	now := time.Now()
	fullContext := make(map[string]interface{})
	for k, v := range ctx {
		fullContext[k] = v
	}
	fullContext["_time"] = map[string]interface{}{
		"hour":      now.Hour(),
		"dayOfWeek": int(now.Weekday()),
		"date":      now.Format("2006-01-02"),
		"timestamp": now.Format(time.RFC3339),
	}
	fullContext["_env"] = map[string]interface{}{
		"name": config.Environment,
	}
	fullContext["_sdk"] = map[string]interface{}{
		"version":  sdkVersion,
		"language": "go",
	}
	if reqVal, ok := ctx["_request"]; ok {
		fullContext["_req"] = reqVal
	}

	var parts []string
	var messages []Message
	var resolvedBlocks []string
	var tools []ToolDefinition
	var errors []string

	// Context schema validation
	if comp.ContextSchema != nil {
		for _, field := range comp.ContextSchema {
			if m, ok := field.(map[string]interface{}); ok {
				required, _ := m["required"].(bool)
				name, _ := m["name"].(string)
				if required && name != "" {
					val := resolveContextPath(fullContext, name)
					if val == nil {
						errors = append(errors, fmt.Sprintf("Required context field missing: '%s'", name))
					}
				}
			}
		}
	}

	// Track roles for multi-message output
	var currentRole string
	var currentRoleContent []string
	var variantID *string

	flushRole := func() {
		if len(currentRoleContent) > 0 && currentRole != "" {
			messages = append(messages, Message{
				Role:    currentRole,
				Content: strings.Join(currentRoleContent, "\n\n"),
			})
			currentRoleContent = nil
		}
	}

	// Visited compositions for cycle detection
	visitedComps := opts.getVisitedCompositions()

	// walk traverses the graph from a given node ID.
	var walk func(nodeID string)
	walk = func(nodeID string) {
		node, ok := nodeMap[nodeID]
		if !ok {
			return
		}

		switch node.Type {
		case "block":
			blockID, _ := node.Data["blockId"].(string)
			block, exists := config.Blocks[blockID]
			if exists {
				content := renderTemplate(block.Content, fullContext)
				parts = append(parts, content)
				resolvedBlocks = append(resolvedBlocks, block.Name)

				// Multi-message support
				blockRole := block.Role
				if blockRole == "" {
					blockRole = "system"
				}
				if currentRole != "" && currentRole != blockRole {
					flushRole()
				}
				currentRole = blockRole
				currentRoleContent = append(currentRoleContent, content)
			}

		case "tool":
			blockID, _ := node.Data["blockId"].(string)
			block, exists := config.Blocks[blockID]
			if exists {
				var inputSchema map[string]interface{}
				if err := json.Unmarshal([]byte(block.Content), &inputSchema); err == nil {
					tools = append(tools, ToolDefinition{
						Name:        block.Name,
						Description: block.Description,
						InputSchema: inputSchema,
					})
				}
				resolvedBlocks = append(resolvedBlocks, block.Name)
			}
			// Falls through to edge-following at bottom

		case "compositionRef":
			compID, _ := node.Data["compositionId"].(string)
			if compID == "" {
				return
			}

			// Find referenced composition
			var refComp *CompositionConfig
			for i := range config.Compositions {
				if config.Compositions[i].ID == compID {
					refComp = &config.Compositions[i]
					break
				}
			}
			if refComp == nil {
				errors = append(errors, fmt.Sprintf("Composition not found for ref: %s", compID))
				return
			}

			// Cycle detection
			if visitedComps[compID] {
				errors = append(errors, fmt.Sprintf("Circular composition reference detected: %s", compID))
				return
			}

			childVisited := make(map[string]bool, len(visitedComps)+1)
			for k, v := range visitedComps {
				childVisited[k] = v
			}
			childVisited[compID] = true

			childResult, err := compose(config, refComp.Name, ctx, &composeOptions{
				visitedCompositions: childVisited,
			})
			if err == nil && childResult.Text != "" {
				parts = append(parts, childResult.Text)
				for _, msg := range childResult.Messages {
					if currentRole != "" && currentRole != msg.Role {
						flushRole()
					}
					currentRole = msg.Role
					currentRoleContent = append(currentRoleContent, msg.Content)
				}
				resolvedBlocks = append(resolvedBlocks, childResult.Blocks...)
				if childResult.VariantID != nil {
					variantID = childResult.VariantID
				}
			}
			return

		case "ifBoolean":
			field, _ := node.Data["field"].(string)
			value := toBool(resolveContextPath(fullContext, field))
			handle := "false"
			if value {
				handle = "true"
			}
			for _, e := range edgesBySource[node.ID] {
				if e.SourceHandle == handle {
					walk(e.Target)
				}
			}
			return

		case "ifSwitch":
			field, _ := node.Data["field"].(string)
			rawValue := resolveContextPath(fullContext, field)
			value := fmt.Sprintf("%v", rawValue)
			cases := toStringSlice(node.Data["cases"])
			match := ""
			if len(cases) > 0 {
				match = cases[len(cases)-1] // default: last case
				for _, c := range cases {
					if c == value {
						match = value
						break
					}
				}
			}
			for _, e := range edgesBySource[node.ID] {
				if e.SourceHandle == match {
					walk(e.Target)
				}
			}
			return

		case "ifPercentage":
			variants := toVariants(node.Data["variants"])
			seed := getSeed(fullContext)
			weights := make([]int, len(variants))
			for i, v := range variants {
				weights[i] = v.weight
			}
			selectedIndex := selectVariant(seed, weights)
			if selectedIndex < len(variants) {
				selectedName := variants[selectedIndex].name
				vID := selectedName
				variantID = &vID
				for _, e := range edgesBySource[node.ID] {
					if e.SourceHandle == selectedName {
						walk(e.Target)
					}
				}
			}
			return

		case "ifExpression":
			expression, _ := node.Data["expression"].(string)
			value := evaluateExpression(expression, fullContext)
			handle := "false"
			if value {
				handle = "true"
			}
			for _, e := range edgesBySource[node.ID] {
				if e.SourceHandle == handle {
					walk(e.Target)
				}
			}
			return
		}

		// For start, merge, output, and any other passthrough nodes: follow all edges
		for _, e := range edgesBySource[node.ID] {
			walk(e.Target)
		}
	}

	// Find start node and begin walking
	for _, n := range nodes {
		if n.Type == "start" {
			walk(n.ID)
			break
		}
	}

	// Flush remaining role content
	flushRole()

	text := strings.Join(parts, "\n\n")

	// Extract model config from composition metadata
	var model string
	var modelCfg *ModelConfig
	if comp.Metadata != nil {
		if mcRaw, ok := comp.Metadata["modelConfig"]; ok {
			if mcMap, ok := mcRaw.(map[string]interface{}); ok {
				if envRaw, ok := mcMap[config.Environment]; ok {
					if envMap, ok := envRaw.(map[string]interface{}); ok {
						if m, ok := envMap["model"].(string); ok {
							model = m
						}
						if model != "" {
							modelCfg = &ModelConfig{}
							if temp, ok := envMap["temperature"].(float64); ok {
								modelCfg.Temperature = &temp
							}
							if mt, ok := envMap["maxTokens"].(float64); ok {
								mt2 := int(mt)
								modelCfg.MaxTokens = &mt2
							}
							if tp, ok := envMap["topP"].(float64); ok {
								modelCfg.TopP = &tp
							}
							if ss, ok := envMap["stopSequences"].([]interface{}); ok {
								for _, s := range ss {
									if sv, ok := s.(string); ok {
										modelCfg.StopSequences = append(modelCfg.StopSequences, sv)
									}
								}
							}
						}
					}
				}
			}
		}
	}

	// Generate assembly ID matching the TS pattern: asm_{timestamp}_{random6}
	id := fmt.Sprintf("asm_%d_%s", time.Now().UnixMilli(), randomString(6))

	return &ComposeResult{
		ID:              id,
		Text:            text,
		Messages:        messages,
		Model:           model,
		Config:          modelCfg,
		Tools:           tools,
		Version:         fmt.Sprintf("v%d", comp.Version),
		VariantID:       variantID,
		TokenCount:      int(math.Round(float64(len(text)) / 4.0)),
		Blocks:          resolvedBlocks,
		CompositionName: compositionName,
		Errors:          errors,
	}, nil
}

// composeOptions holds internal options for recursive composition.
type composeOptions struct {
	visitedCompositions map[string]bool
}

func (o *composeOptions) getVisitedCompositions() map[string]bool {
	if o == nil || o.visitedCompositions == nil {
		return make(map[string]bool)
	}
	return o.visitedCompositions
}

// variant is a helper type for percentage-based routing.
type variant struct {
	name   string
	weight int
}

// toVariants extracts a slice of variants from a data field.
func toVariants(v interface{}) []variant {
	arr, ok := v.([]interface{})
	if !ok {
		return nil
	}
	var result []variant
	for _, item := range arr {
		m, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		name, _ := m["name"].(string)
		weight := toInt(m["weight"])
		result = append(result, variant{name: name, weight: weight})
	}
	return result
}

// toStringSlice extracts a string slice from an interface value.
func toStringSlice(v interface{}) []string {
	arr, ok := v.([]interface{})
	if !ok {
		return nil
	}
	var result []string
	for _, item := range arr {
		s, ok := item.(string)
		if ok {
			result = append(result, s)
		}
	}
	return result
}

// toInt coerces an interface value to int.
func toInt(v interface{}) int {
	switch val := v.(type) {
	case float64:
		return int(val)
	case int:
		return val
	case int64:
		return int(val)
	default:
		return 0
	}
}

// getSeed extracts a seed for percentage routing.
func getSeed(ctx map[string]interface{}) string {
	if req, ok := ctx["_req"]; ok {
		if m, ok := req.(map[string]interface{}); ok {
			if uid, ok := m["userId"].(string); ok && uid != "" {
				return uid
			}
			if sid, ok := m["sessionId"].(string); ok && sid != "" {
				return sid
			}
		}
	}
	return fmt.Sprintf("%d", time.Now().UnixMilli())
}

// randomString generates a random alphanumeric string of the given length.
func randomString(n int) string {
	const charset = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, n)
	for i := range b {
		b[i] = charset[rand.Intn(len(charset))]
	}
	return string(b)
}

File: sdk/src/compose.ts
import type { SDKConfig, ComposeContext, ComposeResult, Message, ToolDefinition, ModelConfig } from "./types"
import { selectVariant } from "./hash"
import { evaluateExpression } from "./expression-parser"
import { renderTemplate } from "./template-engine"
import { SDK_VERSION } from "./version"

export function compose(
  config: SDKConfig,
  compositionName: string,
  context: ComposeContext,
  options?: { sanitize?: boolean; _visitedCompositions?: Set<string> }
): ComposeResult {
  const comp = config.compositions.find((c) => c.name === compositionName)
  if (!comp) throw new Error(`Composition "${compositionName}" not found`)

  const { nodes, edges } = comp.graph
  const nodeMap = new Map(nodes.map((n: any) => [n.id, n]))
  const edgesBySource = new Map<string, any[]>()
  for (const edge of edges) {
    if (!edgesBySource.has(edge.source)) edgesBySource.set(edge.source, [])
    edgesBySource.get(edge.source)!.push(edge)
  }

  const fullContext: Record<string, any> = {
    ...context,
    _time: {
      hour: new Date().getHours(),
      dayOfWeek: new Date().getDay(),
      date: new Date().toISOString().split("T")[0],
      timestamp: new Date().toISOString(),
    },
    _env: { name: config.environment },
    _sdk: { version: SDK_VERSION, language: "typescript" },
  }
  if (context._request) {
    fullContext._req = context._request
  }

  const renderCtx = options?.sanitize ? sanitizeContext(fullContext) : fullContext

  // Validate context schema
  const schemaErrors: string[] = []
  if (comp.contextSchema) {
    for (const field of comp.contextSchema) {
      if (field.required) {
        const parts = field.name.split(".")
        let val: any = renderCtx
        for (const p of parts) {
          val = val?.[p]
        }
        if (val === undefined || val === null) {
          schemaErrors.push(`Required context field missing: '${field.name}'`)
        }
      }
    }
  }

  const parts: string[] = []
  const tools: ToolDefinition[] = []
  const messages: Message[] = []
  const resolvedBlocks: string[] = []
  let variantId: string | null = null

  let currentRole: string | null = null
  let currentRoleContent: string[] = []

  function flushRole() {
    if (currentRoleContent.length > 0 && currentRole) {
      messages.push({ role: currentRole, content: currentRoleContent.join("\n\n") })
      currentRoleContent = []
    }
  }

  function walk(nodeId: string) {
    const node = nodeMap.get(nodeId)
    if (!node) return

    if (node.type === "block") {
      const block = config.blocks[node.data.blockId]
      if (block) {
        const content = renderTemplate(block.content, renderCtx)
        parts.push(content)
        resolvedBlocks.push(block.name)

        const blockRole = block.role || "system"
        if (currentRole !== null && currentRole !== blockRole) {
          flushRole()
        }
        currentRole = blockRole
        currentRoleContent.push(content)
      }
    } else if (node.type === "tool") {
      const block = config.blocks[node.data.blockId]
      if (block) {
        try {
          const inputSchema = JSON.parse(block.content)
          tools.push({ name: block.name, description: block.description ?? "", input_schema: inputSchema })
        } catch {}
        resolvedBlocks.push(block.name)
      }
      // Falls through to edge-following at bottom
    } else if (node.type === "compositionRef") {
      const compositionId = node.data.compositionId as string
      const refComp = config.compositions.find((c) => c.id === compositionId)
      if (!refComp) return

      const visitedComps = options?._visitedCompositions ?? new Set<string>()
      if (visitedComps.has(compositionId)) return // circular ref

      const childVisited = new Set(visitedComps)
      childVisited.add(compositionId)

      const childResult = compose(config, refComp.name, context, {
        sanitize: options?.sanitize,
        _visitedCompositions: childVisited,
      })

      if (childResult.text) {
        parts.push(childResult.text)
        for (const msg of childResult.messages) {
          if (currentRole !== null && currentRole !== msg.role) {
            flushRole()
          }
          currentRole = msg.role
          currentRoleContent.push(msg.content)
        }
      }
      resolvedBlocks.push(...childResult.blocks)
      if (childResult.variantId) variantId = childResult.variantId
      return
    } else if (node.type === "ifBoolean") {
      const value = Boolean(resolve(renderCtx, node.data.field))
      const handle = value ? "true" : "false"
      for (const e of (edgesBySource.get(node.id) ?? []).filter((e: any) => e.sourceHandle === handle)) {
        walk(e.target)
      }
      return
    } else if (node.type === "ifSwitch") {
      const value = String(resolve(renderCtx, node.data.field))
      const cases = node.data.cases ?? []
      const match = cases.includes(value) ? value : cases[cases.length - 1]
      for (const e of (edgesBySource.get(node.id) ?? []).filter((e: any) => e.sourceHandle === match)) {
        walk(e.target)
      }
      return
    } else if (node.type === "ifPercentage") {
      const variants = (node.data.variants as Array<{ name: string; weight: number }>) ?? []
      const seed = renderCtx._req?.userId ?? renderCtx._req?.sessionId ?? String(Date.now())
      const weights = variants.map((v: { name: string; weight: number }) => v.weight)
      const selectedIndex = selectVariant(seed, weights)
      const selectedVariant = variants[selectedIndex]
      if (selectedVariant) {
        variantId = selectedVariant.name
        for (const e of (edgesBySource.get(node.id) ?? []).filter((e: any) => e.sourceHandle === selectedVariant.name)) {
          walk(e.target)
        }
      }
      return
    } else if (node.type === "ifExpression") {
      const expression = (node.data.expression as string) ?? ""
      const value = evaluateExpression(expression, renderCtx)
      const handle = value ? "true" : "false"
      for (const e of (edgesBySource.get(node.id) ?? []).filter((e: any) => e.sourceHandle === handle)) {
        walk(e.target)
      }
      return
    }

    for (const e of edgesBySource.get(node.id) ?? []) {
      walk(e.target)
    }
  }

  const start = nodes.find((n: any) => n.type === "start")
  if (start) walk(start.id)

  // Flush remaining
  flushRole()

  const text = parts.join("\n\n")

  const modelConfig = comp.metadata?.modelConfig?.[config.environment] ?? null
  const model: string | null = modelConfig?.model ?? null
  let configResult: ModelConfig | null = null
  if (model && modelConfig) {
    configResult = {} as ModelConfig
    if (modelConfig.temperature !== undefined) configResult.temperature = modelConfig.temperature
    if (modelConfig.maxTokens !== undefined) configResult.maxTokens = modelConfig.maxTokens
    if (modelConfig.topP !== undefined) configResult.topP = modelConfig.topP
    if (modelConfig.stopSequences !== undefined) configResult.stopSequences = modelConfig.stopSequences
  }

  return {
    id: `asm_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`,
    text,
    messages,
    model,
    config: configResult,
    tools,
    version: `v${comp.version}`,
    variantId,
    tokenCount: Math.round(text.length / 4),
    blocks: resolvedBlocks,
    compositionName,
    errors: schemaErrors,
  }
}

function resolve(ctx: Record<string, any>, path: string): any {
  return path.split(".").reduce((o: any, k: string) => o?.[k], ctx)
}

function sanitizeContext(context: Record<string, any>): Record<string, any> {
  const sanitized: Record<string, any> = {}
  for (const [key, value] of Object.entries(context)) {
    if (typeof value === "string") {
      sanitized[key] = value
        .replace(/\{\{/g, "{ {")
        .replace(/\}\}/g, "} }")
    } else if (value && typeof value === "object" && !Array.isArray(value)) {
      sanitized[key] = sanitizeContext(value)
    } else {
      sanitized[key] = value
    }
  }
  return sanitized
}
